lib/semantics/expression.h

```diff
@@ -100,7 +100,7 @@ struct SetExprHelper {
     if constexpr (ConstraintTrait<T>) {
       Set(x.thing);
     } else {
-      static_assert("bad type");
+      static_assert(static_cast<bool>("bad type"));
     }
```

Comment:
As an observation could this whole section not be written as follows?
```
static_assert(ConstraintTrait<T>, "bad type");
Set(x.thing);
```
The original static assert looks like a mistake as it will never get triggered (a non-0 value is always true!)
I suspect the original should say static_assert(false, "bad type"), in which case the cast wouldn't be necessary.

I realise you didn't write this original code though

Reply:
Thank you David for the reply. As you mentioned that static_assert does not make sense. But the assert when corrected causes compilation to fail (assertion fails). There must be a logical error in the implementation here.

Reply:
I probably meant to type `static_assert(!"bad_type");`.

Reply:
static_assert(!"bad_type"); //should always fail right as it equals static_assert(false)? static_assert(false, "bad_type") should be better as it outputs the message too. But the main problem is this assetion itself is failing.

Reply:
Of course.  Are you also fixing that?

Reply:
I suppose changing the static_assert to dynamic assert is the right way here. If that change is fine, I can do and amend to the pull request.

Reply:
Changing to a runtime `assert`is guaranteed to cause a runtime failure too, since we're definitely calling the function with an invalid T if the static_assert is failing.

Reply:
If the assert is (or should be) failing at compilation time, then there's a problem that needs to be addressed!

Reply:
Yes definitely. I see that types Names and ConstantValue are not being handled and are causing the assert failure. I'll add it as a new issue/PR.